Cross Chain implementation

semaraugusto opened the pull request:
Add tests for cross-chain verification. We can now successfully prove that you're  a member of the group in chain A while using chain B.

drewstone commented on contracts/base/SemaphoreGroups.sol:
```diff
@@ -75,6 +76,35 @@ abstract contract SemaphoreGroups is Context, ISemaphoreGroups {
+	 */
+    function _updateEdge(
+        uint256 _groupId,
+        uint256 _sourceChainID,
```
We can get rid of this for now and parse it from the `src_resource_id`. Refer to https://github.com/webb-tools/protocol-solidity/blob/main/contracts/anchors/LinkableAnchor.sol#L102

drewstone replied:
Please update doc accordingly.

drewstone commented on contracts/interfaces/ISemaphore.sol:
```diff
@@ -42,10 +36,17 @@ interface ISemaphore {
         bytes32 signal,
         uint256 nullifierHash,
         uint256 externalNullifier,
+        // TODO: Create standard encoding for which order each root is supposed to be at.
```
The standard ordering @saiakilesh @semaraugusto is that
- index 0 is _**this**_ anchor's merkle tree root
- index _**i**_>0 is the ordering of the edge list exactly

drewstone reviewed (changes requested):
Seems OK but I'm curious about a few things.
1. Why is TypedChainId present here and passed into the verify proof? What is the point of using typed chain IDs in proof verification for semaphore proofs.
2. The API for updateEdge needs to be updated. Some parameters are no longer needed and naming needs to be changed to match semantically with recent changes in `protocol-solidity`.
3. Why are we passing both `roots` and `root` again? Is this to prove and disclose which root you are a member of? @saiakilesh @semaraugusto  can you please work on a spec of what this implementation is supposed to do entirely so we can match it against implementation. I'm starting to see new changes introduced that deviate from my expectations.

semaraugusto commented:
1. Why is TypedChainId present here and passed into the verify proof? What is the point of using typed chain IDs in proof verification for semaphore proofs.
A: Removed

2. The API for updateEdge needs to be updated. Some parameters are no longer needed and naming needs to be changed to match semantically with recent changes in protocol-solidity.
A: Updated accordingly

3. Why are we passing both roots and root again? Is this to prove and disclose which root you are a member of? @saiakilesh @semaraugusto can you please work on a spec of what this implementation is supposed to do entirely so we can match it against implementation. I'm starting to see new changes introduced that deviate from my expectations.
A: Yes, current implementation is disclosing merkle-root always. This was initially done just to keep the logic that was on the original semaphore circuit. I'll update the spec too
